File: c++/algorithms/insertion/insertion.cpp
#include "insertion.h"


void InsertionSort::sort(int *array, int length) {

	int i, position;

	for (i = 1; i < length; i++) {
		position = i;
		
		while (position > 0 && array[position - 1] > array[i]) {
			position--;
		}

		move(array, i, position);
	}
}

File: js/algorithms/insertion.js

function sortedPosition(array, index) {
    var position = index;
    
    while (position > 0 && array[position - 1] > array[index]) {
        position--;
    }
    
    return position;
}

function sort(array) {
    var count = array.length,
        position,
        value,
        i,
        j
        
    for (i = 1; i < count; i++) {
        position = sortedPosition(array, i);
        value = array[i];

        if (position !== i) {

            for (j = i; j > position; j--) {
                array[j] = array[j - 1];
            }

            array[position] = value;
        }
    }
}

exports.name = 'insertion';
exports.sort = sort;